dao/filtering.go

```diff
@@ -15,46 +15,59 @@ func applyFilters(query *gorm.DB, filters []util.Filter) (*gorm.DB, error) {
 			return nil, fmt.Errorf("failed to parse statement: %v", err)
 		}
 	}
+
 	var filterName string
 	for _, filter := range filters {
-		if query.Statement.Table != "" {
+		// subresource filtering!
+		if filter.Subresource != "" {
+			switch filter.Subresource {
+			case "source_type":
+				query = query.Joins("SourceType")
+				filterName = fmt.Sprintf("%v.%v", `"SourceType"`, filter.Name)
+			case "application_type":
+				query = query.Joins("ApplicationType")
```

Comment:
it looks that it doesn't work when primary collection is source:
```
http://localhost:{{port}}/api/sources/v3.1/sources?filter[application_type][name]=%2Finsights%2Fplatform%2Fcost-management
```
as there is joining table `applications` between `sources` and `application types`

what come to my mind you can use `NewRelationObject` thing used for subcollections: 

```diff
diff --git a/dao/filtering.go b/dao/filtering.go
index 2d071df..a1a41d5 100644
--- a/dao/filtering.go
+++ b/dao/filtering.go
@@ -2,6 +2,7 @@ package dao

 import (
        "fmt"
+       m "github.com/RedHatInsights/sources-api-go/model"
        "strings"

        "github.com/RedHatInsights/sources-api-go/util"
@@ -25,7 +26,15 @@ func applyFilters(query *gorm.DB, filters []util.Filter) (*gorm.DB, error) {
                                query = query.Joins("SourceType")
                                filterName = fmt.Sprintf("%v.%v", `"SourceType"`, filter.Name)
                        case "application_type":
-                               query = query.Joins("ApplicationType")
+                               //query = query.Joins("ApplicationType")
+
+                               relationObject, err := m.NewRelationObject(m.ApplicationType{Id: 3}, 1, DB.Debug())
+                               if err != nil {
+                                       return nil, util.NewErrNotFound("source")
+                               }
+
+                               query = relationObject.HasMany(&m.Source{}, DB.Debug())
+
                                filterName = fmt.Sprintf("%v.%v", `"ApplicationType"`, filter.Name)
                        default:
                                return nil, fmt.Errorf("invalid subresource type [%v]", filter.Subresource)
@@ -38,7 +47,7 @@ func applyFilters(query *gorm.DB, filters []util.Filter) (*gorm.DB, error) {

                switch filter.Operation {
                case "", "eq":
-                       query = query.Where(fmt.Sprintf("%v = ?", filterName), filter.Value[0])
+                       //query = query.Where(fmt.Sprintf("%v = ?", filterName), filter.Value[0])
                case "not_eq":
                        query = query.Where(fmt.Sprintf("%v != ?", filterName), filter.Value[0])
                case "gt":

```

but you need to get ID of application type from in this case name
or change NewRelationObject to work also with `Name`

it is just suggestion maybe it would be doable to write separate query but `New Relation Object` should already cover it.

Reply:
Ah, I should have specified more: the subresources really should only be used for their respective types. Like:

- Source -> filter on SourceType fields
- Application -> filter on ApplicationType fields

like the only URI that I know that does this (minus graphql) is this request from the main platform experience UI: `/api/sources/v3.1/applications?filter%5Bapplication_type%5D%5Bname%5D%5Bcontains%5D=catalog` where they're filtering applications on the application type name. 

---

So it should be ok if we can't filter a Source on the applications connected to it.

Reply:
So we don't run into this later I went and added a check to make sure that the _only_ subresource queries allowed are 
- sources -> source_types 
- applications -> application_types 

that way the user would get a bad request if they ever tried it. Thanks for catching that!

Reply:
thanks @lindgrenj6 !